src/launch/mod.rs

```diff
@@ -192,3 +194,171 @@ impl codicon::Encoder<()> for Measurement {
+
+    /// Userspace address of migration agent. Ignored if MA_EN is 0.
+    pub ma_uaddr: u64,
+
+    /// Indicates if this guest is associated with a migration agent. Otherwise 0.
+    pub ma_en: bool,
+
+    /// Indicates that this launch flow is launching an IMI for the purpose of guest-assisted migration.
+    pub imi_en: bool,
+
+    /// Hypervisor provided value to indicate guest OS visible workarounds.The format is hypervisor defined.
+    pub gosvw: [u8; 16],
+}
+
+/// Encapsulates the various data needed to begin the update process.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Deserialize, Serialize)]
+pub struct SnpUpdate {
+    /// Indicates that this page is part of the IMI of the guest.
+    pub imi_page: bool,
+
+    /// Encoded page type.
+    pub page_type: SnpPageType,
+
+    /// VMPL3 permission mask.
+    pub vmpl3_perms: VmplPerms,
+
+    /// VMPL2 permission mask.
+    pub vmpl2_perms: VmplPerms,
+
+    /// VMPL1 permission mask.
+    pub vmpl1_perms: VmplPerms,
+}
+
+bitflags! {
+    #[derive(Default, Deserialize, Serialize)]
+    /// VMPL permission masks.
+    pub struct VmplPerms: u8 {
+        /// Page is readable by the VMPL.
+        const READ =                1;
+
+        /// Page is writeable by the VMPL.
+        const WRITE =               1 << 1;
+
+        /// Page is executable by the VMPL in CPL3.
+        const EXECUTE_USER =        1 << 2;
+
+        /// Page is executable by the VMPL in CPL2, CPL1, and CPL0.
+        const EXECUTE_SUPERVISOR =  1 << 3;
+    }
+}
+
+/// Encoded page types for a launch update. See Table 58 of the SNP Firmware
+/// specification for further details.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Deserialize, Serialize)]
+pub enum SnpPageType {
+    /// A normal data page.
+    Normal,
+
+    /// A VMSA page.
+    Vmsa,
+
+    /// A page full of zeroes.
+    Zero,
+
+    /// A page that is encrypted but not measured
+    Unmeasured,
+
+    /// A page for the firmware to store secrets for the guest.
+    Secrets,
+
+    /// A page for the hypervisor to provide CPUID function values.
+    Cpuid,
+}
+
+impl SnpPageType {
+    /// Get the encoded value for a page type. See Table 58 of the SNP
+    /// Firmware specification for further details.
+    pub fn value(self) -> u8 {
+        match self {
+            SnpPageType::Normal => 0x1,
+            SnpPageType::Vmsa => 0x2,
+            SnpPageType::Zero => 0x3,
+            SnpPageType::Unmeasured => 0x4,
+            SnpPageType::Secrets => 0x5,
+            SnpPageType::Cpuid => 0x6,
+        }
+    }
+}
+
+/// Encapsulates the data needed to complete a guest launch.
+#[derive(Copy, Clone, Debug, Default, PartialEq, Eq, Deserialize, Serialize)]
+pub struct SnpFinish {
+    /// Userspace address of the ID block. Ignored if ID_BLOCK_EN is 0.
+    pub id_block_uaddr: u64,
+
+    /// Userspace address of the authentication information of the ID block. Ignored if ID_BLOCK_EN is 0.
+    pub id_auth_uaddr: u64,
```

Comment:
These addresses are unsafe. Please use a constructor function that takes a reference to the type and put a phantomdata in the struct so the struct can ensure its lifetime is valid for the lifetime of these addresses.